File: packages/core-ts/src/debouncer.ts
export const Debouncer = {
  MIN_MS: 150,
  INITIAL_MS: 300,
  MAX_MS: 1000,
  SAMPLE_SIZE: 10,

  nextDelay(current: number, latencySamples: number[]): number {
    if (latencySamples.length < this.SAMPLE_SIZE) {
      return clamp(current, this.MIN_MS, this.MAX_MS);
    }
    const slice = latencySamples.slice(-this.SAMPLE_SIZE);
    const avg = Math.round(slice.reduce((a, b) => a + b, 0) / slice.length);
    return Math.max(this.MIN_MS, Math.min(avg, this.MAX_MS));
  },
};

function clamp(n: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, n));
}

File: core/src/main/kotlin/io/autocomplete/engine/Debouncer.kt
package io.autocomplete.engine

import kotlin.math.max
import kotlin.math.min

/**
 * Adaptive debounce delay calculator.
 */
object Debouncer {
    const val MIN_MS = 150L
    const val INITIAL_MS = 300L
    const val MAX_MS = 1000L
    const val SAMPLE_SIZE = 10

    fun nextDelay(
        current: Long,
        latencySamples: List<Long>,
    ): Long {
        if (latencySamples.size < SAMPLE_SIZE) return current.coerceIn(MIN_MS, MAX_MS)
        val avg = latencySamples.takeLast(SAMPLE_SIZE).average().toLong()
        return max(MIN_MS, min(avg, MAX_MS))
    }
}
